rust/xaynet-server/src/storage/api.rs

```diff
@@ -0,0 +1,240 @@
+    ///   `Result::Ok(Option::Some(Vec<(MaskObject, u64)>))`.
+    /// - If two masks exist with the different score, return
+    ///   both in descending order `Result::Ok(Option::Some(Vec<(MaskObject, u64)>))`.
+    async fn get_best_masks(&mut self) -> StorageResult<Option<Vec<(MaskObject, u64)>>>;
+
+    /// Gets the number of unique masks.
+    async fn get_number_of_unique_masks(&mut self) -> StorageResult<u64>;
+
+    /// Deletes all coordinator data. This includes the coordinator
+    /// state as well as the [`SumDict`], [`SeedDict`] and `mask` dictionary.
+    async fn delete_coordinator_data(&mut self) -> StorageResult<()>;
+
+    /// Deletes the [`SumDict`], [`SeedDict`] and `mask` dictionary.
+    async fn delete_dicts(&mut self) -> StorageResult<()>;
+
+    /// Sets the latest global model id.
+    ///
+    /// # Behavior
+    ///
+    /// - If no global model id has been set yet, set and return the new id `Result::Ok(String))`.
+    /// - If the global model id already exists, override and return the new id `Result::Ok(String))`.
+    async fn set_latest_global_model_id(&mut self, id: &str) -> StorageResult<()>;
+
+    /// Gets the latest global model id.
+    ///
+    /// # Behavior
+    ///
+    /// - If the global model id does not exist, return `Result::Ok(None)`.
+    /// - If the global model id exists, return `Result::Ok(String))`.
+    async fn get_latest_global_model_id(&mut self) -> StorageResult<Option<String>>;
+}
+
+#[async_trait]
+/// An abstract model storage.
+pub trait ModelStorage
+where
+    Self: Clone + Send + Sync + 'static,
+{
+    /// Sets a global model.
+    ///
+    /// # Behavior
+    ///
+    /// - If the global model already exists (has the same model id), return `Result::Err(StorageError))`.
+    /// - If the global model does not exist, set the model and return the corresponding id.
+    async fn set_global_model(&mut self, id: &str, global_model: &Model) -> StorageResult<()>;
+
+    /// Gets a global model.
+    ///
+    /// # Behavior
+    ///
+    /// - If the global model does not exist, return `Result::Ok(Option::None)`.
+    /// - If the global model exists, return `Result::Ok(Option::Some(Model))`.
+    async fn get_global_model(&mut self, id: &str) -> StorageResult<Option<Model>>;
+
+    /// Creates a unique global model id by using the round id and the round seed in which
+    /// the global model was created.
+    ///
+    /// The format of the default implementation is as follows `roundid_roundseed`,
+    /// where the [`RoundSeed`] is encoded in hexadecimal.
+    fn create_global_model_id(round_id: u64, round_seed: &RoundSeed) -> String {
+        let round_seed = hex::encode(round_seed.as_slice());
+        format!("{}_{}", round_id, round_seed)
+    }
+}
+
+/// A wrapper that contains the result of the "add sum participant" operation.
+#[derive(Deref)]
+pub struct SumPartAdd(pub(crate) Result<(), SumPartAddError>);
+
+impl SumPartAdd {
+    /// Unwraps the result.
```

Comment:
actually this unwraps the wrapper, but not the result. not sure what would be a good terminology here though.

```suggestion
    /// Unwraps the wrapper.
```

same for the other unwraps below.

Reply:
Hm in [std](https://doc.rust-lang.org/std/io/struct.IntoInnerError.html?search=intoinner) they also have both versions. What about `Unwraps this wrapper, returning the underlying result.`?

Reply:
sounds good 👍